\chapter{Deep Learning \& AI}

\section{}


\subsubsection*{Generative Adversarial Networks}

Read \cite{goodfellow2014generative}




\section{ML Finance Project}

\subsubsection*{\href{https://stackabuse.com/time-series-prediction-using-lstm-with-pytorch-in-python/}{example w/ multivariate time series in PyTorch}}


\begin{quest}
\item
	\cloze
	Neural networks can be constructed using the \pyth{torch.nn} package.

\item
	Import the package for constructing neural networks in PyTorch.
	\begin{ans}
		\pyth{import torch.nn as nn}
	\end{ans}

\item \cloze Seaborn comes with built-in datasets.

\item Load seaborn's flights dataset.
	\begin{ans}
		\pyth{flight_data = sns.load_dataset("flights")}
	\end{ans}

\item
	Why must time series data be scaled for sequence predictions?
	\begin{ans}
		When a network is fit on unscaled data, it is possible for large inputs to slow down the learning and convergence of your network and in some cases prevent the network from effectively learning your problem.
	\end{ans}

\item
	sklearn import for scaling data?
	\begin{ans}
		\pyth{from sklearn.preprocessing import MinMaxScaler}
	\end{ans}
\end{quest}





\begin{quote}
We know the field is fast moving. If the reader looking for more recent free reading resources, there are some good introductory/tutorial/survey papers on Arxiv; I happen to be compiling a list of them.
\end{quote}

 One of said review papers \cite{raghu2020survey}

\cite[hello]{raghu2020survey}

\section{Deep Learning for Genomic Risk Scores}

\begin{quotation}
``
A central aim of computational genomics is to identify variants (SNPs) in the genome which increase risks for diseases. Current analyses apply linear regression to identify SNPs with large associations, which are collected into a function called a Polygenic Risk Score (PRS) to predict disease for newly genotyped individuals. This project is broadly interested in whether we can improve performance of genomic risk scores using modern machine learning techniques.

A recent study assessed the disease prediction performance of neural networks in comparison to conventional PRSs, but did not find evidence of improvement. This project will explore whether neural networks can improve performance by incorporating gene expression data to the training process. Gene expression is often integrated with SNP data in Transcriptome-Wide Association Studies (TWAS), which bear some resemblance to neural network architectures with SNPs as input nodes, genes as intermediate nodes, and disease status as the output node. Modeling this process as a neural network however will require defining a more unconventional architecture in which a small subset of hidden nodes is anchored to observed values.

This project is designed for students with experience in machine learning topics and preferably with deep learning tools such as tensorflow or pytorch. Students should also be interested in applying machine learning and statistics to genomics applications.'' - Jie Yuan
\end{quotation}

\textbf{Terms to know}: Computational genomics, variants, single-nucleotide polymorphism (SNP), genome, Polygenic Risk Score (PRS), Transcriptome-Wide Association Studies (TWAS), gene(s), genomics


\subsection{Polygenic Risk Scores (paper) \cite{wray2010multi}}
\subsubsection*{Abstract (mining)}
\begin{description}
\item[recurrence risks] :
	In genetics, the likelihood that a hereditary trait or disorder present in one family member will occur again in other family members\footnote{\url{https://www.cancer.gov/publications/dictionaries/genetics-dictionary/def/recurrence-risk}}.

	``Evidence for genetic contribution to complex diseases is described by recurrence risks to relatives of diseased individuals.''

	This is distinguished from recurrence risk for cancer, which is the chance that a cancer that has been treated will recur.

\item[genome] :
	An organism’s complete set of DNA, including all of its genes. Each genome contains all of the information needed to build and maintain that organism. In humans, a copy of the entire genome—more than 3 billion DNA base pairs—is contained in all cells that have a nucleus \footnote{\url{https://ghr.nlm.nih.gov/primer/hgp/genome}}.

	``genome-wide association''

\item[allosome] :
	(1) A sex chromosome such as the X and Y human sex chromosomes. (2) An atypical chromosome \footnote{\url{https://www.merriam-webster.com/medical/allosome}}.

	allo- (Greek): other, differnt

\item[autosome] :
	Any chromosome that is not a sex chromosome. The numbered chromosomes.

	auto (Greek): self, one's own, by oneself, of oneself

	-some, soma (Greek): body

\item[allele] :
	(genetics) One of a number of alternative forms of the same gene occupying a given position, or locus, on a chromosome.

	Borrowed from German Allel, shortened from English allelomorph. Ultimately from the Ancient Greek prefix allēl- from állos (“other”).

	``their effects and allele frequencies''

	allelomorph: another term for allele.

\item[risk loci] :

	``genome-wide association studies allow a description of the genetics of the same diseases in terms of risk loci...''

\item[monozygotic] :

	``monozygotic twins''

\item[empirical] :

	``generate results more consistent with empirical estimates''

	\item[genetic variants]:
\end{description}



\subsection{Neural Networks for Genomic Prediction (paper) \cite{pinto2019can}}

\subsection{Transcriptome Wide Association}

